File: Leetcode/October/Day2_Combination_Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector <int> r;
        vector <vector<int>> res;
        findNumbers(candidates, res, r, target, 0);
        return res;
    }
    void findNumbers(vector<int>& arr, vector<vector<int> >& res, vector<int>& r, int sum, int i)
    {
        if(sum < 0) // if sum becomes negative
            return;
        if(sum == 0){  // if sum becomes zero
            res.push_back(r);
            return;
        }
            
        while(i < arr.size() && sum - arr[i] >= 0)  // if i less than arr size and sum - arr[i] >= 0
        {
            r.push_back(arr[i]); // add them to the list
            findNumbers(arr, res, r, sum-arr[i], i);
            i++;
            
            r.pop_back(); // backtracking
            
        }
        
    }
};

File: Leetcode/Monthly Challenges/October_LeetCoding_Challenge/Day2_Combination_Sum/Day2_Combination_Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
 	  List<List<Integer>> res = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        helper(candidates, 0, target, 0, temp, res);
        return res;

    }

    void helper(int[] candidates, int start, int target, int sum, List<Integer> list, List<List<Integer>> result) {
        if (sum > target)
            return;
        if (sum == target) {
            result.add(new ArrayList<>(list));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            list.add(candidates[i]);
            helper(candidates, i, target, sum + candidates[i], list, result);
            list.remove(list.size() - 1);
        }       
    }
}
